commit aa36d711e945e65fa87410927800f01878a8faed
Author: Sagi Grimberg <sagi@grimberg.me>
Date:   Sun Nov 13 13:24:20 2022 +0200

    nvme-auth: convert dhchap_auth_list to an array
    
    We know exactly how many dhchap contexts we will need, there is no need
    to hold a list that we need to protect with a mutex. Convert to
    a dynamically allocated array. And dhchap_context access state is
    maintained by the chap itself.
    
    Make dhchap_auth_mutex protect only the ctrl host_key and ctrl_key
    in a fine-grained lock such that there is no long lasting acquisition
    of the lock and no need to take/release this lock when flushing
    authentication works.
    
    Signed-off-by: Sagi Grimberg <sagi@grimberg.me>
    Signed-off-by: Christoph Hellwig <hch@lst.de>

diff --git a/drivers/nvme/host/nvme.h b/drivers/nvme/host/nvme.h
index e431ec348d52..ef23c6c6e2a3 100644
--- a/drivers/nvme/host/nvme.h
+++ b/drivers/nvme/host/nvme.h
@@ -337,8 +337,8 @@ struct nvme_ctrl {
 
 #ifdef CONFIG_NVME_AUTH
 	struct work_struct dhchap_auth_work;
-	struct list_head dhchap_auth_list;
 	struct mutex dhchap_auth_mutex;
+	struct nvme_dhchap_queue_context *dhchap_ctxs;
 	struct nvme_dhchap_key *host_key;
 	struct nvme_dhchap_key *ctrl_key;
 	u16 transaction;